sys/clif/clif.c

```diff
@@ -0,0 +1,378 @@
+        return CLIF_NO_SPACE;
+    }
+
+    buf[pos++] = LF_PATH_BEGIN_C;
+
+    memcpy(&buf[pos], target, target_len);
+    pos += target_len;
+
+    buf[pos++] = LF_PATH_END_C;
+
+    return pos;
+}
+
+ssize_t clif_add_link_separator(char *buf, size_t maxlen)
+{
+    if (!buf) {
+        return 1;
+    }
+
+    if (maxlen < 1) {
+        return CLIF_NO_SPACE;
+    }
+
+    *buf = LF_LINK_SEPARATOR_C;
+    return 1;
+}
+
+ssize_t clif_add_param(clif_param_t *param, char *buf, size_t maxlen)
+{
+    assert(param && param->key);
+
+    param->key_len = strlen(param->key);
+    /* count param name size and separator ';' */
+    size_t req_space = param->key_len + 1;
+    size_t pos = 0;
+    int quoted = strcmp(param->key, LF_PARAM_SIZE) ? 1 : 0;
+
+    if (param->value) {
+        if (!param->value_len) {
+            param->value_len = strlen(param->value);
+        }
+        /* count also '=' */
+        req_space += param->value_len +  1;
+    }
+
+    if (quoted) {
+        req_space += 2;
+    }
+
+    if (!buf) {
+        return req_space;
+    }
+
+    if (req_space > maxlen) {
+        return CLIF_NO_SPACE;
+    }
+
+    /* add parameter separator ';' */
+    buf[pos++] = LF_PARAM_SEPARATOR_C;
+
+    /* add parameter name */
+    memcpy(&buf[pos], param->key, param->key_len);
+    pos += param->key_len;
+
+    /* add parameter value if defined */
+    if (param->value) {
+        buf[pos++] = LF_PARAM_ATTR_SEPARATOR_C;
+
+        if (quoted) {
+            buf[pos++] = '"';
+        }
+
+        memcpy(&buf[pos], param->value, param->value_len);
+        pos += param->value_len;
+
+        if (quoted) {
+            buf[pos++] = '"';
+        }
+    }
+
+    return pos;
+}
+
+ssize_t clif_get_link(const char *input, size_t input_len, char **output)
+{
+    assert(input);
+    char *link_end;
+
+    ssize_t path_size = clif_get_target(input, input_len, output);
+    if (path_size < 0) {
+        DEBUG("Path not found\n");
+        return CLIF_NOT_FOUND;
+    }
+
+    (*output)--;
+    /* look for link separator ',' */
+    link_end = memchr(*output, LF_LINK_SEPARATOR_C,
+                      (input_len - (*output - input)));
+
+    if (!link_end) {
+        /* there is no other link */
+        return (input_len - (*output - input));
+    }
+    else {
+        return (link_end - *output);
+    }
+}
+
+ssize_t clif_get_target(const char *input, size_t input_len, char **output)
+{
+    assert(input);
+    char *target_end;
+
+    *output = memchr(input, LF_PATH_BEGIN_C, input_len);
+    if (!*output) {
+        DEBUG("Path start not found\n");
+        return CLIF_NOT_FOUND;
+    }
+    *output += 1;
+
+    target_end = memchr(*output, LF_PATH_END_C, (input + input_len) - *output);
+    if (!target_end) {
+        DEBUG("Path end not found\n");
+        return CLIF_NOT_FOUND;
+    }
+    ssize_t res = target_end - *output;
+    return res;
+}
+
+ssize_t clif_get_param(char *input, size_t input_len, clif_param_t *param)
```

Comment:
Does it make sense to require a non-null param? This would be more consistent with the other 'clif_get...' functions. It also would save six 'if param' conditionals. For `clif_decode_link()` you could just pass a dummy param.

Reply:
> This would be more consistent with the other 'clif_get...' functions

Makes sense. Changed it.